lib/aclgenerator.py

```diff
@@ -280,32 +254,100 @@ def __init__(self, pol, exp_info):
         for term in terms:
           # Only verify optional keywords if the term is active on the platform.
           err = []
-          if term.platform:
-            if self._PLATFORM not in term.platform:
-              continue
-          if term.platform_exclude:
-            if self._PLATFORM in term.platform_exclude:
-              continue
           for el, val in term.__dict__.items():
             # Private attributes do not need to be valid keywords.
-            if (val and el not in self._valid_keywords
-                and not el.startswith('flatten')):
-              err.append(el)
+            if (val and el not in supported_tokens and not
+                el.startswith('flatten')):
+                err.append(el)
+            # ignore Liskov's rule.
+            if (val and type(val) is list and
+                el in supported_sub_tokens):
+                ns = set(val) - supported_sub_tokens[el]
+                # hack support for ArbitraryOptions in junos. todo, add the
+                # junos options into the lexer, then we can nuke .*
+                # shenanigans.
+                if ns and '.*' not in supported_sub_tokens[el]:
+                  err.append(' '.join(ns))
           if err:
-            raise UnsupportedFilterError(
-                'Unsupported optional keyword(s): %s in term %s of policy %s '
-                'for platform %s' % (' '.join(err), term.name,
-                                     header.FilterName(self._PLATFORM),
-                                     self._PLATFORM))
+            all_err.append(
+              '%s contains unsupported keywords (%s) for target %s in policy %s'
+              % (term.name, ' '.join(err), self._PLATFORM, pol.filename))
         continue
-
+    if all_err:
+      raise UnsupportedFilterError('\n %s' % '\n'.join(all_err))
     self._TranslatePolicy(pol, exp_info)
 
   def _TranslatePolicy(self, pol, exp_info):
     # pylint: disable=unused-argument
     """Translate policy contents to platform specific data structures."""
     raise Error('%s does not implement _TranslatePolicies()' % self._PLATFORM)
 
+  def _buildTokens(self):
+    """provide a default for supported tokens and sub tokens.
+
+    Returns:
+      tuple
+    """
+    # Set of supported keywords for a given platform.  Values should be in
+    # undercase form, eg, icmp_type (not icmp-type)
+    supported_tokens = {'action',
```

Comment:
I tend to prefer literals over builtins, but the real reason i used it was from this:

`In [1]: from dis import dis

In [2]: dis('{1}')
  1           0 LOAD_CONST               0 (1)
              3 BUILD_SET                1
              6 RETURN_VALUE

In [3]: dis('set([1])')
  1           0 LOAD_NAME                0 (set)
              3 LOAD_CONST               0 (1)
              6 BUILD_LIST               1
              9 CALL_FUNCTION            1 (1 positional, 0 keyword pair)
             12 RETURN_VALUE
`

using the set() call just makes more calls under the hood. The other reason is that it makes set comps easier as well:

`In [4]: bar=[1,2,3,3,4,4,5]

In [5]: {foo for foo in bar}
Out[5]: {1, 2, 3, 4, 5}`